### PR TITLE
29: Add JWT for login, register and refresh token

### PR DESCRIPTION
**Pull request by jchiavaro**
Added json web tokens for securely transmitting information between front and backend:
https://jwt.io
Now, after the users login or register, the backend returns a token that will be used by the frontend in every request to the api.
A new endpoint was added to allow users to renew the token: `/api/user/token?token="..."`
Also, express-jwt was added to allow routes to use auth, or exclude the ones that don't need to be accessed by an authorized user.

### CODE REVIEW

File: server/routes/user/refreshToken.route.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  var token = req.body.token || req.query.token || req.headers['x-access-token'] || req.headers['authorization'];
+  if (!token) {
+    return res.status(401).json({message: 'Invalid Token'});
+  }
+  jwt.verify(token, process.env.JWT_SECRET, function(err, user) {
+    if (err) throw err;
+    UserModel.findById({
+      '_id': user._id
+    }, function(err, usr) {
+      if (err) throw err;
```
Comment (matiassequeira) [inline server/routes/user/refreshToken.route.js]: This error is unhandled?

File: server/helper/passwordHasher.js [hunk 2]
```diff
@@ -9,7 +10,32 @@ function hashPassword(passwordToHash){
...
+	    admin: user.admin,
+      _id: user._id.toString()
+    }
+
+    return jwt.sign(u, process.env.JWT_SECRET, {
+    	expiresIn: 60 * 60 * 24
+    });
+}
+
+function getUserForToken(user) {
```
Comment (matiassequeira) [inline server/helper/passwordHasher.js]: This function is unused?

**Comment by matiassequeira** [comment]
Everything seems good, I only left two questions as comments. Once you answer them I merge the PR

File: server/helper/passwordHasher.js (hunk 2 above)
Comment (jchiavaro) [reply server/helper/passwordHasher.js]: Right, I’ll remove it.

Comment (jchiavaro) [reply server/helper/passwordHasher.js]: Right, I’ll remove it.

File: server/routes/user/refreshToken.route.js (hunk 1 above)
Comment (jchiavaro) [reply server/routes/user/refreshToken.route.js]: Good catch, I’ll return 500 instead. Thanks

**Comment by jchiavaro** [comment]
Ok, I updated the requested changes

**Pull request merged**